### PR TITLE
Feature social login

### PR DESCRIPTION
**Pull request by rwhite27**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -18,6 +18,8 @@
         "dmkit/phalcon-jwt-auth": "dev-master",
         "duncan3dc/fork-helper": "^2.2.0",
         "firebase/php-jwt": "^5.0",
+        "hybridauth/hybridauth": "dev-master",
```
Comment (gabbanaesteban) [inline composer.json]: Do not use dev-master. Never. @rwhite27
If someone introduces a breaking change, you fucked.

File: src/Providers/SocialLoginProvider.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+        $container->setShared(
+            'socialLogin',
+            function () use ($config) {
+
+                /**
+                 * @todo Change the way provider information is handled
+                 */
+                $providers = [
+                    'callback'=> 'https://apidev.kanvas.dev/v1/users/social',
```
Comment (gabbanaesteban) [inline src/Providers/SocialLoginProvider.php]: what about using .env for that magic link?

File: src/Api/Controllers/AuthController.php [hunk 3]
```diff
@@ -142,4 +145,32 @@ public function changeUserEmail(string $hash): Response
...
+    {
+        $accessToken = $this->request->getPost('access_token','string');
+        $providerName =  $this->request->getPost('provider','string');
+
+        $source = Sources::findFirst([
+            'conditions'=>'title = ?0 and is_deleted = 0',
+            'bind'=>[$providerName]
+        ]);
+
+        if (!is_object($source)) {
```
Comment (gabbanaesteban) [inline src/Api/Controllers/AuthController.php]: FindFirst method returns false if no record was found.
it is faster using and cleaner using !$source instead.

File: src/Traits/SocialLoginTrait.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        /**
+         * Verify if user exists
+         */
+        $user = Users::findFirst([
+            'conditions' => 'email = ?0 and is_deleted = 0',
+            'bind' => [$userProfile->email]
+        ]);
+
+        if (!is_object($user) && !is_object($userLinkedSource)) // User does not exist and has not been linked to a source
```
Comment (gabbanaesteban) [inline src/Traits/SocialLoginTrait.php]: Do not make if statements that long. It is harder to read for a dev.
What about flipping the statement?

if ($user && $userLinkedSource){
	return $this->loginSocial($user);
}

// Here you don't need the else anymore.

// Registering the user

File: src/Providers/SocialLoginProvider.php [hunk 5]
```diff
@@ -26,7 +26,7 @@ function () use ($config) {
                 $providers = [
                     'callback'=> '',
                     'providers'=>[
-                        'Facebook'=> ['enabled'=> true, 'keys'=>['id'=>'1197131063788181','secret'=>'1ee2138ace609549f524945ab9f3db3d']]
```
Comment (gabbanaesteban) [inline src/Providers/SocialLoginProvider.php]: You have to do a git rm or change credentials...
Remember that this project could be public once and there are bots that search for those secret keys and tokens...

File: src/Traits/SocialLoginTrait.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        {
+            $random = new Random();
+            $password = '123456';
```
Comment (kaioken) [inline src/Traits/SocialLoginTrait.php]: @rwhite27 change password to something random

File: composer.json (hunk 1 above)
Comment (rwhite27) [reply composer.json]: We are going to user 'dev-master' for now. Cant seem to find an stable version for version 3.

Comment (gabbanaesteban) [reply composer.json]: What about using a pre-release tag?
https://github.com/hybridauth/hybridauth/tree/v3.0.0-rc.10
It is better than nothing.
This is the rc @rwhite27

File: src/Traits/SocialLoginTrait.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+        /**
+         * Lets Login or Signup the user
+         */
+        $userProfile = current($data);
+
+        /**
+        * Lets find if user has a linked source by social network id
+        */
+        $userLinkedSource =  UserLinkedSources::findFirst([
```
Comment (kaioken) [inline src/Traits/SocialLoginTrait.php]: @rwhite27 no need to find the users look for the relationship taht should be good enough

File: src/Traits/SocialLoginTrait.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ */
+trait SocialLoginTrait
+{
+    /**
+     * Facebook Login
+     * @param Sources $source
+     * @param string $accessToken
+     * @return Array
+     */
+    protected function facebookLogin(Sources $source, string $accessToken): array
```
Comment (kaioken) [inline src/Traits/SocialLoginTrait.php]: @rwhite27 if the trait is named social login why name the function facebook login? isnt it better just facebook?

File: src/Traits/SocialLoginTrait.php [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            'conditions'=>'source_id = ?0 and source_users_id_text = ?1 and is_deleted = 0',
+            'bind'=>[
+                    $source->id,
+                    $userProfile->identifier
+                ]
+        ]);
+
+        if ($userLinkedSource->getUser()) {
+            $facebookAdapter->disconnect();
+            return $this->loginSocial($user);
```
Comment (kaioken) [inline src/Traits/SocialLoginTrait.php]: @rwhite27 this trait depends from this function thats on another trait, if so defined it as a abstract method to force the user to have it

File: src/Api/Controllers/CompaniesController.php [hunk 10]
```diff
@@ -148,4 +131,25 @@ public function delete($id): Response
...
+
+    /**
+     * Format output.
+     *
+     * @param [type] $results
+     * @return void
+     */
+    protected function processOutput($results)
+    {
+        foreach ($results as $key => $value) {
```
Comment (kaioken) [inline src/Api/Controllers/CompaniesController.php]: @rwhite27 document why are we doing this

**Pull request merged**